File: src/main/java/komu/blunt/types/TypeCheckResult.java
package komu.blunt.types;

import static java.util.Collections.emptyList;

import java.util.List;

public final class TypeCheckResult<T> {

    public final List<Predicate> predicates;
    public final T value;
    
    public TypeCheckResult(List<Predicate> predicates, T value) {
        this.predicates = predicates;
        this.value = value;
    }
    
    public TypeCheckResult(T value) {
        this.predicates = emptyList();
        this.value = value;
    }

    @Override
    public String toString() {
        return predicates + " => " + value;
    }
}

File: src/main/kotlin/komu/blunt/types/checker/TypeCheckResult.kt
package komu.blunt.types.checker

import komu.blunt.types.Predicate
import java.util.Collections.emptyList

data class TypeCheckResult<out T>(val value: T, val predicates: List<Predicate> = emptyList()) {

    operator fun plus(predicates: List<Predicate>): TypeCheckResult<T> =
        TypeCheckResult(value, predicates + this.predicates)

    override fun toString() = "$predicates => $value"
}
